// tailwindcss/lib/lib/expandApplyAtRules.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return expandApplyAtRules;
    }
});
const _postcss = /*#__PURE__*/ _interop_require_default(require("postcss"));
const _postcssselectorparser = /*#__PURE__*/ _interop_require_default(require("postcss-selector-parser"));
const _generateRules = require("./generateRules");
const _escapeClassName = /*#__PURE__*/ _interop_require_default(require("../util/escapeClassName"));
const _applyImportantSelector = require("../util/applyImportantSelector");
const _pseudoElements = require("../util/pseudoElements");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/** @typedef {Map<string, [any, import('postcss').Rule[]]>} ApplyCache */ function extractClasses(node) {
    /** @type {Map<string, Set<string>>} */ let groups = new Map();
    let container = _postcss.default.root({
        nodes: [
            node.clone()
        ]
    });
    container.walkRules((rule)=>{
        (0, _postcssselectorparser.default)((selectors)=>{
            selectors.walkClasses((classSelector)=>{
                let parentSelector = classSelector.parent.toString();
                let classes = groups.get(parentSelector);
                if (!classes) {
                    groups.set(parentSelector, classes = new Set());
                }
                classes.add(classSelector.value);
            });
        }).processSync(rule.selector);
    });
    let normalizedGroups = Array.from(groups.values(), (classes)=>Array.from(classes));
    let classes = normalizedGroups.flat();
    return Object.assign(classes, {
        groups: normalizedGroups
    });
}
let selectorExtractor = (0, _postcssselectorparser.default)();
/**
 * @param {string} ruleSelectors
 */ function extractSelectors(ruleSelectors) {
    return selectorExtractor.astSync(ruleSelectors);
}
function extractBaseCandidates(candidates, separator) {
    let baseClasses = new Set();
    for (let candidate of candidates){
        baseClasses.add(candidate.split(separator).pop());
    }
    return Array.from(baseClasses);
}
function prefix(context, selector) {
    let prefix = context.tailwindConfig.prefix;
    return typeof prefix === "function" ? prefix(selector) : prefix + selector;
}
function* pathToRoot(node) {
    yield node;
    while(node.parent){
        yield node.parent;
        node = node.parent;
    }
}
/**
 * Only clone the node itself and not its children
 *
 * @param {*} node
 * @param {*} overrides
 * @returns
 */ function shallowClone(node, overrides = {}) {
    let children = node.nodes;
    node.nodes = [];
    let tmp = node.clone(overrides);
    node.nodes = children;
    return tmp;
}
/**
 * Clone just the nodes all the way to the top that are required to represent
 * this singular rule in the tree.
 *
 * For example, if we have CSS like this:
 * ```css
 * @media (min-width: 768px) {
 *   @supports (display: grid) {
 *     .foo {
 *       display: grid;
 *       grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
 *     }
 *   }
 *
 *   @supports (backdrop-filter: blur(1px)) {
 *     .bar {
 *       backdrop-filter: blur(1px);
 *     }
 *   }
 *
 *   .baz {
 *     color: orange;
 *   }
 * }
 * ```
 *
 * And we're cloning `.bar` it'll return a cloned version of what's required for just that single node:
 *
 * ```css
 * @media (min-width: 768px) {
 *   @supports (backdrop-filter: blur(1px)) {
 *     .bar {
 *       backdrop-filter: blur(1px);
 *     }
 *   }
 * }
 * ```
 *
 * @param {import('postcss').Node} node
 */ function nestedClone(node) {
    for (let parent of pathToRoot(node)){
        if (node === parent) {
            continue;
        }
        if (parent.type === "root") {
            break;
        }
        node = shallowClone(parent, {
            nodes: [
                node
            ]
        });
    }
    return node;
}
/**
 * @param {import('postcss').Root} root
 */ function buildLocalApplyCache(root, context) {
    /** @type {ApplyCache} */ let cache = new Map();
    root.walkRules((rule)=>{
        // Ignore rules generated by Tailwind
        for (let node of pathToRoot(rule)){
            var _node_raws_tailwind;
            if (((_node_raws_tailwind = node.raws.tailwind) === null || _node_raws_tailwind === void 0 ? void 0 : _node_raws_tailwind.layer) !== undefined) {
                return;
            }
        }
        // Clone what's required to represent this singular rule in the tree
        let container = nestedClone(rule);
        let sort = context.offsets.create("user");
        for (let className of extractClasses(rule)){
            let list = cache.get(className) || [];
            cache.set(className, list);
            list.push([
                {
                    layer: "user",
                    sort,
                    important: false
                },
                container
            ]);
        }
    });
    return cache;
}
/**
 * @returns {ApplyCache}
 */ function buildApplyCache(applyCandidates, context) {
    for (let candidate of applyCandidates){
        if (context.notClassCache.has(candidate) || context.applyClassCache.has(candidate)) {
            continue;
        }
        if (context.classCache.has(candidate)) {
            context.applyClassCache.set(candidate, context.classCache.get(candidate).map(([meta, rule])=>[
                    meta,
                    rule.clone()
                ]));
            continue;
        }
        let matches = Array.from((0, _generateRules.resolveMatches)(candidate, context));
        if (matches.length === 0) {
            context.notClassCache.add(candidate);
            continue;
        }
        context.applyClassCache.set(candidate, matches);
    }
    return context.applyClassCache;
}
/**
 * Build a cache only when it's first used
 *
 * @param {() => ApplyCache} buildCacheFn
 * @returns {ApplyCache}
 */ function lazyCache(buildCacheFn) {
    let cache = null;
    return {
        get: (name)=>{
            cache = cache || buildCacheFn();
            return cache.get(name);
        },
        has: (name)=>{
            cache = cache || buildCacheFn();
            return cache.has(name);
        }
    };
}
/**
 * Take a series of multiple caches and merge
 * them so they act like one large cache
 *
 * @param {ApplyCache[]} caches
 * @returns {ApplyCache}
 */ function combineCaches(caches) {
    return {
        get: (name)=>caches.flatMap((cache)=>cache.get(name) || []),
        has: (name)=>caches.some((cache)=>cache.has(name))
    };
}
function extractApplyCandidates(params) {
    let candidates = params.split(/[\s\t\n]+/g);
    if (candidates[candidates.length - 1] === "!important") {
        return [
            candidates.slice(0, -1),
            true
        ];
    }
    return [
        candidates,
        false
    ];
}
function processApply(root, context, localCache) {
    let applyCandidates = new Set();
    // Collect all @apply rules and candidates
    let applies = [];
    root.walkAtRules("apply", (rule)=>{
        let [candidates] = extractApplyCandidates(rule.params);
        for (let util of candidates){
            applyCandidates.add(util);
        }
        applies.push(rule);
    });
    // Start the @apply process if we have rules with @apply in them
    if (applies.length === 0) {
        return;
    }
    // Fill up some caches!
    let applyClassCache = combineCaches([
        localCache,
        buildApplyCache(applyCandidates, context)
    ]);
    /**
   * When we have an apply like this:
   *
   * .abc {
   *    @apply hover:font-bold;
   * }
   *
   * What we essentially will do is resolve to this:
   *
   * .abc {
   *    @apply .hover\:font-bold:hover {
   *      font-weight: 500;
   *    }
   * }
   *
   * Notice that the to-be-applied class is `.hover\:font-bold:hover` and that the utility candidate was `hover:font-bold`.
   * What happens in this function is that we prepend a `.` and escape the candidate.
   * This will result in `.hover\:font-bold`
   * Which means that we can replace `.hover\:font-bold` with `.abc` in `.hover\:font-bold:hover` resulting in `.abc:hover`
   *
   * @param {string} selector
   * @param {string} utilitySelectors
   * @param {string} candidate
   */ function replaceSelector(selector, utilitySelectors, candidate) {
        let selectorList = extractSelectors(selector);
        let utilitySelectorsList = extractSelectors(utilitySelectors);
        let candidateList = extractSelectors(`.${(0, _escapeClassName.default)(candidate)}`);
        let candidateClass = candidateList.nodes[0].nodes[0];
        selectorList.each((sel)=>{
            /** @type {Set<import('postcss-selector-parser').Selector>} */ let replaced = new Set();
            utilitySelectorsList.each((utilitySelector)=>{
                let hasReplaced = false;
                utilitySelector = utilitySelector.clone();
                utilitySelector.walkClasses((node)=>{
                    if (node.value !== candidateClass.value) {
                        return;
                    }
                    // Don't replace multiple instances of the same class
                    // This is theoretically correct but only partially
                    // We'd need to generate every possible permutation of the replacement
                    // For example with `.foo + .foo { Б─╕ }` and `section { @apply foo; }`
                    // We'd need to generate all of these:
                    // - `.foo + .foo`
                    // - `.foo + section`
                    // - `section + .foo`
                    // - `section + section`
                    if (hasReplaced) {
                        return;
                    }
                    // Since you can only `@apply` class names this is sufficient
                    // We want to replace the matched class name with the selector the user is using
                    // Ex: Replace `.text-blue-500` with `.foo.bar:is(.something-cool)`
                    node.replaceWith(...sel.nodes.map((node)=>node.clone()));
                    // Record that we did something and we want to use this new selector
                    replaced.add(utilitySelector);
                    hasReplaced = true;
                });
            });
            // Sort tag names before class names (but only sort each group (separated by a combinator)
            // separately and not in total)
            // This happens when replacing `.bar` in `.foo.bar` with a tag like `section`
            for (let sel of replaced){
                let groups = [
                    []
                ];
                for (let node of sel.nodes){
                    if (node.type === "combinator") {
                        groups.push(node);
                        groups.push([]);
                    } else {
                        let last = groups[groups.length - 1];
                        last.push(node);
                    }
                }
                sel.nodes = [];
                for (let group of groups){
                    if (Array.isArray(group)) {
                        group.sort((a, b)=>{
                            if (a.type === "tag" && b.type === "class") {
                                return -1;
                            } else if (a.type === "class" && b.type === "tag") {
                                return 1;
                            } else if (a.type === "class" && b.type === "pseudo" && b.value.startsWith("::")) {
                                return -1;
                            } else if (a.type === "pseudo" && a.value.startsWith("::") && b.type === "class") {
                                return 1;
                            }
                            return 0;
                        });
                    }
                    sel.nodes = sel.nodes.concat(group);
                }
            }
            sel.replaceWith(...replaced);
        });
        return selectorList.toString();
    }
    let perParentApplies = new Map();
    // Collect all apply candidates and their rules
    for (let apply of applies){
        let [candidates] = perParentApplies.get(apply.parent) || [
            [],
            apply.source
        ];
        perParentApplies.set(apply.parent, [
            candidates,
            apply.source
        ]);
        let [applyCandidates, important] = extractApplyCandidates(apply.params);
        if (apply.parent.type === "atrule") {
            if (apply.parent.name === "screen") {
                let screenType = apply.parent.params;
                throw apply.error(`@apply is not supported within nested at-rules like @screen. We suggest you write this as @apply ${applyCandidates.map((c)=>`${screenType}:${c}`).join(" ")} instead.`);
            }
            throw apply.error(`@apply is not supported within nested at-rules like @${apply.parent.name}. You can fix this by un-nesting @${apply.parent.name}.`);
        }
        for (let applyCandidate of applyCandidates){
            if ([
                prefix(context, "group"),
                prefix(context, "peer")
            ].includes(applyCandidate)) {
                // TODO: Link to specific documentation page with error code.
                throw apply.error(`@apply should not be used with the '${applyCandidate}' utility`);
            }
            if (!applyClassCache.has(applyCandidate)) {
                throw apply.error(`The \`${applyCandidate}\` class does not exist. If \`${applyCandidate}\` is a custom class, make sure it is defined within a \`@layer\` directive.`);
            }
            let rules = applyClassCache.get(applyCandidate);
            // Verify that we can apply the class
            for (let [, rule] of rules){
                if (rule.type === "atrule") {
                    continue;
                }
                rule.walkRules(()=>{
                    throw apply.error([
                        `The \`${applyCandidate}\` class cannot be used with \`@apply\` because \`@apply\` does not currently support nested CSS.`,
                        "Rewrite the selector without nesting or configure the `tailwindcss/nesting` plugin:",
                        "https://tailwindcss.com/docs/using-with-preprocessors#nesting"
                    ].join("\n"));
                });
            }
            candidates.push([
                applyCandidate,
                important,
                rules
            ]);
        }
    }
    for (let [parent, [candidates, atApplySource]] of perParentApplies){
        let siblings = [];
        for (let [applyCandidate, important, rules] of candidates){
            let potentialApplyCandidates = [
                applyCandidate,
                ...extractBaseCandidates([
                    applyCandidate
                ], context.tailwindConfig.separator)
            ];
            for (let [meta, node] of rules){
                let parentClasses = extractClasses(parent);
                let nodeClasses = extractClasses(node);
                // When we encounter a rule like `.dark .a, .b { Б─╕ }` we only want to be left with `[.dark, .a]` if the base applyCandidate is `.a` or with `[.b]` if the base applyCandidate is `.b`
                // So we've split them into groups
                nodeClasses = nodeClasses.groups.filter((classList)=>classList.some((className)=>potentialApplyCandidates.includes(className))).flat();
                // Add base utility classes from the @apply node to the list of
                // classes to check whether it intersects and therefore results in a
                // circular dependency or not.
                //
                // E.g.:
                // .foo {
                //   @apply hover:a; // This applies "a" but with a modifier
                // }
                //
                // We only have to do that with base classes of the `node`, not of the `parent`
                // E.g.:
                // .hover\:foo {
                //   @apply bar;
                // }
                // .bar {
                //   @apply foo;
                // }
                //
                // This should not result in a circular dependency because we are
          H┐Л(ХЦЪЪЪH┐Х   H┐3юH┐д(цлллллллH█╜5  цллллллллллллллллллллH┐Л(ХёЪЪЪH┐ХрЪЪЪH┐3юH┐д(цлллллллллллллллллллH┐Л(┘рt9┐Йt(┐Йt┐Зt
╦   H┐д(цХ  КХш  ╤юH┐д(цI▀пH┐д(И   M┘ю∙аH┐д(И$  ллллллллH┴\$H┴t$H┴|$ AVH┐Л H▀РL▀Я3иХ√  └ю└х   ХИ  ┼ь┬D$@@╥┐=]:   ┘е   гM:     Х@  └юtOХ▀  Хф	  ХИ	  H█З  H█
с  Хп
  ┘юu)Ху  └юt H█╡  H█
{  Х╓
  гЬ9     @2Ъ┼кХ╡  @└Ъu?Х╪	  H▀ьH┐8 t$H▀хХВ  └юtL▀ф╨   I▀нH▀L▀
&  AЪяЪ-4  ╦   К3юH▀\$0H▀t$8H▀|$HH┐д A^ц╧   Х─	  ░лллллллH┴\$WH┐Л @┼Ы▀И3  ┘ю~SЪх┴щ3  Хь  ┼ь┬D$8┐=O9  uDХ  ХЦ  Хн
  ┐%79   Х2  ┼кХО  3р@┼оХ  └юt╦   К3юH▀\$0H┐д _ц╧   ХВ  ░ллллллллллH▀дH┴X L┴@┴PH┴HVWAVH┐Л@I▀П▀ЗL▀Я┘рu9D3  3юИП   █BЪ┐ЬwEH▀╪  H┘юu
гD$0   КЪ  ▀ь┴D$0┘ю└╢   L▀ф▀вI▀нХ▄ЩЪЪ▀ь┴D$0┘ю└≥   L▀ф▀вI▀нХ║ЮЪЪ▀ь┴D$0┐Ъu8┘юu4L▀ф3рI▀нХ┘ЮЪЪL▀ф3рI▀нХHЩЪЪH▀A  H┘юtL▀ф3рI▀нЪ▌  ┘Ъt┐Ъu@L▀ф▀вI▀нХЩЪЪ▀ь┴D$0┘юt)H▀  H┘юu	█X┴\$0КL▀ф▀вI▀нЪK  ▀ь┴D$0К3ш┴\$0▀цH▀\$xH┐д@A^_^цлллллллллллллH┴\$H┴t$WH┐Л I▀Ь▀зH▀Я┐ЗuХk  L▀г▀сH▀нH▀\$0H▀t$8H┐д _И┐ЧЪЪлллллллллff└     H;
≥1  РuHааfВаЪЪРuРцHаиИo   ллллллл@SH┐Л H▀ыЪ!  ╧   ┴7  Х  3иЪ9  H▀кЪ8  ┐=7   u
╧   ХА  Ъъ  H▀х╨	 юH┐д [HЪ%Ё  лллллллллллH┴L$H┐Л8╧   Ъ╪  ┘юt╧   м)H█
з1  Х╜   H▀D$8H┴а2  H█D$8H┐юH┴Q2  H▀╙2  H┴1  H▀D$@H┴2  гУ0  	 югО0     гЫ0     ╦   Hkю H█
Я0  Hг   ╦   Hkю H▀
i0  H┴L ╦   HkюH▀
\0  H┴L H█
╗  ХкЧЪЪH┐д8цлллллл@SVWH┐Л@H▀ыЪ  H▀ЁЬ   3ЪE3юH█T$`H▀нЪМ  H┘юt9H┐d$8 H█L$hH▀T$`L▀хH┴L$0L▀фH█L$pH┴L$(3иH┴\$ Ъф  Ъг┐Ъ|╠H┐д@_^[цлллллллH┐Л(ХС  ┘юt!eH▀%0   H▀HКH;хt3юПH╠
\5  uН2юH┐д(ц╟КВлллллллH┐Л(ХЁ  ┘юtХ  КХ≈  ▀хХ2	  ┘юt2юКХI	  ╟H┐д(цллллллллH┐Л(3иХi  └ю∙юH┐д(цлллллллH┐Л(ХO
  └юu2юКХB
  └юuХ9
  КЛ╟H┐д(цллллллллH┐Л(Х
  Х
  ╟H┐д(цлллллллH┴\$H┴l$H┴t$WH┐Л I▀ЫI▀П▀зH▀ИХТ  ┘юu┐ШuL▀ф3рH▀мH▀гЪн  H▀T$X▀L$PH▀\$0H▀l$8H▀t$@H┐д _И╒  ллллллллH┐Л(Х╖  ┘юtH█
04  H┐д(ИA  Х▌	  ┘юuХ  H┐д(цллллллллH┐Л(3иХa	  H┐д(ИX	  лллллллл@SH┐Л ╤4  ┘и╩   Dц┬4  Х╙  Х)	  └юu2юКХ	  └юu	3иХ	  КЙ┼цH┐д [цллллллл@SH┐Л ─=ц3   ▀ы┘┬   ┐Ы┤┤   ХИ  ┘юt(┘шu$H█
n3  Х⌡  ┘юuH█
v3  Х▀  ┘юtK2юКPH▀z-  ╧@   ▀б┐Ю?+хH┐хЪHсхH3бH┴-3  H┴.3  H┴/3  H┴03  H┴13  H┴23  ф33  ╟H┐д [ц╧   Хщ  лллллллллH┐ЛL▀а╦MZ  f9e⌡ЪЪuxHc
≤⌡ЪЪH█U⌡ЪЪHй│9PE  u_╦  f9AuTL+б╥AH█QHп╥AH█─L█йH┴$I;яt▀JL;аr
▀BаL;юrH┐б(Къ3рH┘рu2юК┐z$ }2юК
╟К2юК2юH┐дцлллллллл@SH┐Л ┼ыХ╖  3р┘юt└шuH┤"2  H┐д [цлллллллл@SH┐Л ─=@2   ┼ыt└рuХb  ┼кХ[  ╟H┐д [цллллллл@SH┐Л H▀,  H▀ы▀йH3с1  ┐А?HсйH┐ЗЪu
H▀кХм  КH▀сH█
Ё1  ХЬ  3и┘юHDкH▀аH┐д [цлллллллллH┐Л(Х÷ЪЪЪHВьюВьЪхH┐д(цлллллллллИО  ллллллллллллллл@SH┐Л H█Ш	  H▀ыH┴ЖбtХц  H▀цH┐д [цллллллH┴\$ UH▀ЛH┐Л H▀P+  H╩2╒ъ-≥+  H;цutH┐e H█MЪz  H▀EH┴EЪд  ▀юH1EЪю  ▀юH█M H1EЪь  ▀E H█MHаЮ H3E H3EH3аH╧ЪЪЪЪЪЪ  H#аH╧3╒ъ-≥+  H;цHDаH┴м*  H▀\$HHВпH┴ф*  H┐д ]цллллллллH█
╧0  HЪ%р
  ллллллH█
╔0  И2  ллллллллH┐Л(ХкТЪЪH┐ХЗТЪЪH┐H┐д(цлллллллллH█2  цлллллллл┐%m0   цллллллллH┴\$UH█╛$@ШЪЪH│Лю  ▀ы╧   ЪБ
  ┘юt▀км)╧   Х╪ЪЪЪ3рH█MПA╦п  ХI  H█MПЪщ
  H▀²Х   H█∙ь  H▀кE3юЪ╩
  H┘юt<H┐d$8 H██Ю  H▀∙ь  L▀хH┴L$0L▀цH██Х  H┴L$(H█MПH┴L$ 3иЪ┼
  H▀┘х  H█L$PH┴┘Х   3рH█┘х  A╦≤   H┐юH┴┘┬   Х╡  H▀┘х  H┴D$`гD$P  @гD$T   ЪЧ  ┐ЬH█D$PH┴D$@H█EП■цH┴D$H3иЪ
	  H█L$@Ъ
	  ┘юu└шu█HХ╤ЧЪЪH▀°$п  H│дю  ]цлллллллллH┴\$WH┐Л H█©  H█=╦  КH▀H┘юtЪЬ
  H┐цH;ъrИH▀\$0H┐д _цллллллллH┴\$WH┐Л H█▀  H█=└  КH▀H┘юtЪ╢
  H┐цH;ъrИH▀\$0H┐д _цллллллллллллллллб  лллллллллH┴\$H┴t$WH┐Л3югM(     3иг=(     ╒D▀а3ЪD▀кA│ПntelA│ЯGenuD▀р▀П3и█GEх╒A│РineI┴$Eй┴\$D▀ы┴L$┴T$uPH┐
Ь'  Ъ%П?Ъ=ю t(=` t!=p t╟ЫЭЪ┐Ь w$H╧     HёаsD▀и-  A┐хD┴╬-  КD▀╣-  ╦   ;П|&3и╒┴$▀Ш┴\$┴L$┴T$╨Ц	sA┐хD┴├-  A╨Цspгa'     г['     A╨ЦsUA╨ЦsN3ипHаБ HпH┴T$ H▀D$ $<u2▀+'  ┐хг'     ┴'  @Жг t┐х г'     ┴Ъ&  H▀\$(3юH▀t$0H┐д_цллллллллл╦   цлллллл3ю9э&  ∙юцллллллллИщ   лллллллллллллллллЪ%°  ллллллЪ%┬  ллллллЪ%d  ллллллЪ%P  ллллллЪ%<  ллллллЪ%(  ллллллЪ%  ллллллЪ%≤  ллллллЪ%■  ллллллЪ%░  ллллллЪ%▄  ллллллЪ%░  ллллллЪ%▄  ллллллЪ%┬  ллллллЪ%└  ллллллЪ%─  ллллллЪ%|  ллллллЪ%─  ллллллЪ%|  ллллллЪ%х  ллллллллH┴\$H┴l$H┴t$WH┐Л0I▀ыI▀ЬH▀РH▀ИХоОЪЪH┴\$(L▀оH┐d$  L▀фH▀уH▀H┐иХЪЪЪH▀\$@┐иЪH▀l$H┘юH▀t$PHаH┐д0_цллллллH┴\$H┴l$H┴t$WH┐Л@I▀ыI▀ЬH▀РH▀ИХgОЪЪL▀T$pL▀оL┴T$0L▀фH┐d$( H▀уH▀H┴\$ ХёЧЪЪH▀\$P┐иЪH▀l$X┘юH▀t$`HаH┐д@_цлллллллл╟цллллллллл3юцлллллллллH┐Л(M▀A8H▀йI▀яХ   ╦   H┐д(цллллллл@SE▀H▀зA┐ЦЬL▀иAЖ L▀яtA▀@McPВьLяHcхL#яIcцJ▀H▀C▀HH▀CЖDt╤D┐ЮПLхL3йI▀и[И≥РЪЪлллллллллH▀дH┴XH┴hH┴pH┴x AVH┐Л I▀Y8H▀РM▀ПH▀ИI▀яH▀нI▀ЫL█CХdЪЪЪ▀E$fЖь╦   EюAВьDюD┘CtL▀оM▀фH▀жH▀мХJЩЪЪH▀\$0H▀l$8H▀t$@H▀|$HH┐д A^цлллллллff└     H┐ЛL┴$L┴\$M3шL█T$L+пMBсeL▀%   M;сРsfA│Б ПM█⌡ ПЪЪAф M;сРuОL▀$L▀\$H┐дРцллллллЪ%3  ллллллЪ%Г  лллллллллллллff└     ЪЮллллллH█┼H   ИdпЪЪллллллH█┼P   И▌╪ЪЪлллллл@UH┐Л H▀Й┼M@H┐д ]ИfЖЪЪллллллл@UH┐Л H▀ЙХ█ТЪЪ┼M8H┐д ]ИDЖЪЪллллллл@UH┐Л0H▀ЙH▀▀H┴L$(┴T$ L█
gМЪЪL▀Ep▀UhH▀M`Х╚СЪЪ░H┐д0]цллллллл@UH▀ЙH▀3и│8  ю■а▀а]цллллллллллллллллллH┐Л(H▀
е!  H┘иtХ⌠мЪЪH┐д(цллллллллллллллллллллллH┴\$H┴t$WH┐Л H█=z)  H▀7К'H▀чH▀vHЪ  D  L▀ц3рH▀хHЪС  D  H┘ЖuтH┴7H█▐)  H┐гH;Ьu╪H▀\$0H▀t$8H┐д _цл                     П▒ ─   ░▓ ─   ╗x ─   f ─   0 ─    ─   `! ─   ═  ─                                                                                         ═▒ ─                   хs ─   пs ─   Xt ─   6        u                                                                                                                                    Д┴      ┴      ж┴              ┴      └┬      °┼      ─┼              \┴              x┬      j┴      ╟┴              2▀              0┼      ┼      ┼              v┴      ╗┬      б┴      L┬              З┬              н┼      т┬       ┴      ╨┼              Б┼              Ч┼              R┼      >┼      l┼              J┴      2┴      Й┬      ╬┬      ■┬      b┬      ┬┴      Ь┴              ▀              д├      Б├      Н├      
┤      ┤      (┤      D┤      h┤      ├┤      ╕┤      ф┤      h█      э┤      Ф┤      ╓├      ┌├      d├      J├      4├      R█              ├      Ь┘              Ч┤              @i ─   ─n ─            ─   @ ─   ` ─   ─ ─   ю ─                   п[ ─    \ ─   @\ ─                                              @   `   ─   ю      0   Ю   `   ╟   п   ═    `!   ─"   ═"    7   ─8    9   ╟?  п?   pA   ░B   ЮB   пD    F   @F   ─G   ПH   ╟L   ЮN   ЮO   @P    Q   `R   ╟R   ЮR    S    S   @S   ─S   юS   ╟U   юU   PW   п[    \   @\   p\   ╟_   ░`   f   @i   Po   ─o                             Unknown exception       Exception       ReturnHr    LogHr       FailFast        % h s ( % u ) \ % h s ! % p :           % h s ! % p :           ( c a l l e r :   % p )         
